### ISSUE TITLE
Introduce StateFlow

### ISSUE THREAD

**Issue by elizarov** [Kotlin/kotlinx.coroutines#1973]
We need to be able to conveniently use `Flow` to represent an updateable state in applications. This change introduces `StateFlow` -- an updateable value that represents a state and _is_ a `Flow`. The design is flexible to fit a variety of needs:

* `StateFlow<T>` interface is a _read-only_ view that gives access to the current `value` and implements a `Flow<T>` to collect updates to the values. 
* `MutabaleStateFlow<T>` interface adds value-modification operation.

A `MutableStateFlow(x)` constructor function is provided. It returns an implementation of `MutableStateFlow` with the given initial value. It can be exposed to the outside world as either `StateFlow<T>` if fast non-reactive access to the `value` is needed, or as `Flow<T>` if only reactive view of updates to the value is needed.

Core state flow API can be summarized like this:

```kotlin
package kotlinx.coroutines.flow

interface StateFlow<T> : Flow<T> {
    val value: T // always availabe, reading it never fails
}

interface MutableStateFlow<T> : StateFlow<T> {
    override var value: T // can read & write value
}

fun <T> MutableStateFlow(value: T): MutableStateFlow<T> // constructor fun
```

Implementation is available in PR #1974.

 ### StateFlow vs ConflatedBroadcastChannel

Conceptually state flow is similar to `ConflatedBroadcastChannel`  and is designed to completely replace `ConflatedBroadcastChannel` in the future. It has the following important improvements:

* `StateFlow` is simpler because it does not have to implement all the `Channel` APIs, which allows for faster, garbage-free implementation, unlike `ConflatedBroadcastChannel` implementation that allocates objects on each emitted value.
* `StateFlow` always has a value that can be safely read at any time via `value` property. Unlike `ConflatedBroadcastChannel`, there is no way to create a state flow without a value.
* `StateFlow` has a clear separation into a read-only `StateFlow` interface and a `MutableStateFlow`.
* `StateFlow` conflation is based on equality, unlike conflation in `ConflatedBroadcastChannel` that is based on reference identity. It is a stronger, more practical conflation policy, that avoids extra updates when data classes are emitted. You can consider it to have an embedded `distinctUntilChanged` out-of-the-box.
* `StateFlow` cannot be currently closed like `ConflatedBroadcastChannel` and can never represent a failure. This feature might be added in the future if enough compelling use-cases are found.

`StateFlow` is designed to better cover typical use-cases of keeping track of state changes in time, taking more pragmatic design choices for the sake of convenience.

### Example

For example, the following class encapsulates an integer state and increments its value on each call to `inc`:

```kotlin
class CounterModel {
    private val _counter = MutableStateFlow(0) // private mutable state flow
    val counter: StateFlow<Int> get() = _counter // publicly exposed as read-only state flow

    fun inc() {
        _counter.value++ // convenient state update
    }
}
```

### Experimental status

The initial version of this design is going to be introduced under `@ExperimentalCoroutinesApi`, but it is highly unlikely that the core of the design as described above, is going to change. It is expected to be stabilized quite fast. 

There are also some future possible enhancements (see below) that are not provided at this moment.

### Possible future enhancement: Closing state flow

A state flow could be optionally `closed` in a similar way to channels. When state flow is closed all its collectors complete normally or with the specified exception. Closing a state flow transitions it to the terminal state. Once the state flow is closed its value cannot be changed.  The most recent value of the closed state flow is still available via `value` property. Closing a state is appropriate in situations where the source of the state updates is permanently destroyed.

To support closing there would be a `MutableStateFlow.close(cause: Throwable? = null): Boolean` method for the owner of the state flow to close it via its writeable reference and `StateFlow.isClosed: Boolean` property on read-only interface. 

### Possible future enhancement: Turning Flow into a StateFlow
 
A regular `Flow` is cold. It does not have the concept of the last value and it only becomes active when collected. We could introduce a `stateIn` operator to turn any `Flow` into a hot `StateFlow` in the future as a part of the family of flow-sharing operators (see #1261). It could be designed to become a replacement for `broadcastIn` operator. It would subsume the need of having to have a separate "state flow builder" as you can simply write `flow { .... }.stateIn(scope)` to launch a coroutine that emits the values according to the code in curly braces.

**Comment by Dico200** [comment]
I like it. You mentioned users can use convenient syntax like `state.value++` - does this mean the implementation will not be thread safe?

**Comment by Dico200** [comment]
Perhaps a confusing question - atomicity is something that can be added as a layer around the api `if (compareAndSet(old, new)) state.value = new`. But I would just like to ask what you're planning for the state machine and whether it will be threadsafe like `Job`, or more light and less thread safe.

**Comment by elizarov** [comment]
The internal state machine is fully thread-state, but its public API is primarily designed for cases when there is a single "owner" that sequentially updates the state's value, so we don't provide operations like `compareAndSet`. They would only complicate API. We don't have compelling use-cases on the table to support them and, if need, you can always use other mechanisms to coordinate updaters if you have many of them.

**Comment by Dico200** [comment]
Sounds good! 🎉

**Comment by zach-klippenstein** [comment]
This API looks great, really exciting!

It would be useful to have a version of `stateIn` that takes an initial value and does not suspend. This would make it easy to define a "loading" or default state while waiting for a resource to spin up.

```kotlin
fun <T> Flow<T>.stateIn(initialValue: T, scope: CoroutineScope): StateFlow<T>
```

Some example use cases:

- A streaming GRPC call that won't emit until it gets a value from the network, but you can provide a reasonable initial state locally.
- A settings store that will emit whenever a setting is changed, but allows you to provide a default value that will be used if the setting has never been set or until it's initially loaded from disk.

This function could be implemented on top of the existing `stateIn` using `onStart` and `runBlocking` (on the JVM target) but it would be more efficient to just initialize the internal `MutableStateFlow` eagerly.

**Comment by erikc5000** [comment]
Having the `StateFlow` interface represent a read-only state flow while the constructor function with the same name is actually mutable seems like it could lead to a lot of state flows getting unintentionally exposed as mutable. Is there a reason to not use `MutableStateFlow` as the name of the constructor function?

Otherwise, it's looking pretty good to me!

**Comment by igorwojda** [comment]
I also think that we should have separate `MutableStateFlow()` constructor to avoid confusion and keep things consistent (List, Map, Android LiveData all of them have Mutable equivalents)

It would be good to add extension to simplify conversion from MutableStateFlow to StateFlow
```
fun <T> MutableStateFlow<T>.toStateFlow() = this as StateFlow<T>

class CounterModel {
	private val _counter = MutableStateFlow(0)
	val counter = _counter.toStateFlow()
}
```

**Comment by JakeWharton** [comment]
That would be an "as", not a "to", but I'm opposed to that. I don't see it adding meaningful value compared to just specifying the type explicitly. It's actually more characters, not that character count provides any meaningful metric. Plus when you enable library mode (explicit types always required) it means that function wouldn't even be needed.

**Comment by ZakTaccardi** [comment]
> Unlike ConflatedBroadcastChannel, there is no way to create a state flow without a value.

I definitely feel weird about this - what if the initial state is built asynchronously? I currently use an actor coroutine to build state.

```kotlin
private val stateChannel = ConflatedBroadcastChannel<ProfileId?>()

private val actor = actor<Intention> {
  stateChannel.send(getActiveProfileId())
  
  for (intention in channel) {
    // update state channel based on incoming intentions
  }
}
```

I guess the new pattern would be:

```kotlin
val intentionsChannel = Channel<Intention>()
val states = flow<ProfileId?> {
    emit(getActiveProfileId())
  
    for (intention in intentionsChannel) {
        // emit new state based on incoming intentions
    }
}
  .stateIn(scope = this)
```

But this leaves the `Channel<Intention>` used to communicate into the `StateFlow<T>`  disconnected, even though they are related.

**Comment by ZakTaccardi** [comment]
> You can consider it to have an embedded distinctUntilChanged out-of-the-box.

Will this be configurable? Could I inject my own predicate to evaluate if two items are diffirent?

**Comment by elizarov** [comment]
There is no configurability for comparisons that are used to avoid sending extra state-changes in the initial release. We are trying to keep it as simple as basic `Set` operations are. Create a separate issue with use-cases for configurable behaviour, please, if you have some use-cases for it in mind.

**Comment by igorwojda** [comment]
Right `asStateFlow() `would be better name indeed

Regarding the value well - main one is ease of use - this gives us much faster code completion, more convenient usage with IDE. 

When it comes to library mode we know that majority of devs are writing apps, not libraries, but still these two solution are not mutually exclusive.

**Comment by fvasco** [comment]
A valuable implementation should provide -at least- the defensive programming, otherwise a plain cast can be enough.

```kotlin
fun <T> MutableStateFlow<T>.asStateFlow() = object : StateFlow<T> by this {}
```

**Comment by elizarov** [comment]
@erikc5000 @igorwojda Indeed, the naming for constructor functions is quite a controversial issue here. We have two basic naming options:

* `MutableStateFlow()` is totally explicit but somewhat extraneous. Unlike collections, where it does make sense to construct both a `MutableList` and a read-only `List`, here we have a case when creating a read-only version itself does not make any use. 

* `StateFlow()` (as proposed in this issue) is shorter, but has a downside that the actual return type is `MutableStableFlow`. We followed a precedent of `Job()` constructor here, that is in a similar situation and returns a `CompletableJob` that the owner of the Job needs while exposing a `Job` to the outside world. There were no complaints so far on it, but it is worth noting that it is not that widely use as `StateFlow` is expected to become. Also, note that `CompletableDeferred()` constructor naming does not follow this precedent.

As for "accidentally exposing a mutable type" it does not seem to be a problem if you are writing small, self-contained code. On the contrary, it is consistent with Kotlin's "public default" policy that is aimed at reducing verbosity for non-library end-user application code: 

```
// never state the type explicitly
val myState = StateFlow(initialValue) 
// now I can update my state
myState.value = another
// and I can operate on it as flow
myState.collect { .... }
// the fact that mystate: MutableStateFlow is quite secondary here
```

However, when you do write a library or a larger application, then you have to be careful at the boundaries with respect to the types you expose. The problem I see here is that a proposed declaration pattern here reads somewhat weird:

1) This proposal

```
private val _counter = StateFlow(0)
val counter: StateFlow<Int> get() = _counter 
```

It reads as if you created a `StateFlow` and then cast it to `StateFlow`, which looks weird unless you know the details of the API or IDE is helping you with inferred types. Adding `asStateFlow()` looks useful, for cases where you are not in the library (so you don't want to be explicit about), but when you want to control types between different layers of your application and don't want to explicitly spell the full type:

Adding `asStateFlow()`:

```
private val _counter = StateFlow(0)
val counter get() = _counter.asStateFlow()
```

This somewhat weird look of `StateFlow as StateFlow` remains here, though, just as before.

2) Changing name of the constructor to `MutableStateFlow`:

```
private val _counter = MutableStateFlow(0)
val counter: StateFlow<Int> get() = _counter.asStateFlow() // A
val counter get() = _counter.asStateFlow() // or B
```

Either way, with `MutableStateFlow(value)` it becomes more explicit at the expense of longer and harder-to-discover name. See, the feature will be known as `StateFlow`, so people will be looking for `StateFlow` to use it. That's also the reason as to why most documentation is concentrated in KDoc for `StateFlow` interface. 

All in all, I don't have a strong opinion on my own here and cannot strongly convince myself in either direction. What does this community think? Let's do a quick poll:

👍 For `StateFlow(value)` constructor.
🚀 For `MutableStableFlow(value)` constructor.

**Comment by LouisCAD** [comment]
It is possible to use the `MutableStableFlow(value)` constructor, and have a deprecated (error level) for `StateFlow(value)` with a `ReplaceWith` clause to use the proper and explicit one.

**Comment by fvasco** [comment]
`StateFlow(value)` and `MutableStableFlow(value)` look like `listOf(value)` and `mutableListOf(value)`.

**Comment by RobertKeazor** [comment]
Very Nice Work! The API actually looks very similar to LiveData AND MutableLiveData. But def like the very simple design .

**Comment by erikc5000** [comment]
```kotlin
class CounterModel {
   private val _counter = MutableStateFlow(0)
   val counter: StateFlow<Int> get() = _counter
}
```

...is basically the pattern that's used with Android's `LiveData` type, so it should look familiar to a lot of people.

**Comment by zach-klippenstein** [comment]
Not all kotlin devs are Android devs 😉

I like @LouisCAD's suggestion about adding an error-deprecated alias function for discoverability. I don't really think the extra length/explicitness of the name `MutableStateFlow` is a problem - it's much more clear, and code is read a lot more than it's written. 

That said, as another alternative, how important is it to make this function look like a constructor? If it were called `stateFlow()`, it would still be fairly discoverable, but since it doesn't look like a constructor it wouldn't imply quite as strongly that the return type is exactly `StateFlow`.

**Comment by elizarov** [comment]
**UPDATE**: Based on your feedback we've decided to rename `StateFlow()` constructor to `MutableStableFlow()` and to fast-track all the core `StateFlow` APIs to stabilization. However, we'll pull out and postpone all the questionable parts of state flow design, including "Closing of the flow" and `stateIn` operator. The main reason here is that `stateIn` is only one operator of a big family of "sharing operators" (see #1261 for big chunk of the discussion). Releasing just `shareIn`, even as a "preview" feature, carries a high risk that people would start cramming all of their flow-sharing use-cases into this single narrowly-scoped operator, while there will a whole family of them, covering a variety of use-cases. I'll keep you in the loop on the updates to the design process for sharing. It is the next big priority for the team. I'll update PR and introductory text to this issue soon.

**Comment by ZakTaccardi** [comment]
As `StateFlow<T>` must always have a value, I'm still curious how we are supposed to leverage use of a `StateFlow<T>` that does not have a value already initialized 

See https://github.com/kotlin/kotlinx.coroutines/issues/1973#issuecomment-621521467

**Comment by nschwermann** [comment]
@ZakTaccardi make a sealed class that has a representation for not being initialized

**Comment by ZakTaccardi** [comment]
> make a sealed class that has a representation for not being initialized
@nschwermann 

this is a non-starter. I want to cleanly support the scenario where my `Flow<T>` behaves like a `StateFlow<T>` but doesn't emit a value until it's been lazily initialized. And I would prefer if this use case is supported by the Coroutines framework rather than a custom wrapper .

This sounds like something a `.stateIn(..)` operator could potentially address, which is why I'm looking forward to understanding if this is something the Coroutines framework will officially support in some way

**Comment by nschwermann** [comment]
@ZakTaccardi in that case I would use the sealed class that I suggested and then the filter operator.

**Comment by ZakTaccardi** [comment]
> in that case I would use the sealed class that I suggested and then the filter operator.

Yes, this works. But it's a workaround. I would then have to create a custom `LazyMutableStateFlow<T> : MutableStateFlow<T>` wrapper.

But I'm looking for an idiomatic approach

**Comment by zach-klippenstein** [comment]
@ZakTaccardi How would `value` behave before the first value has been emitted then? Would there be a way to query the `StateFlow` to find out if it has a value yet?

**Comment by ZakTaccardi** [comment]
@zach-klippenstein 

> How would value behave before the first value has been emitted then? Would there be a way to query the StateFlow to find out if it has a value yet?

If you are referring to the `LazyMutableStateFlow<T>` I referred to in https://github.com/Kotlin/kotlinx.coroutines/issues/1973#issuecomment-622100466, it would be the same as `value` and `valueOrNull` from `ConflatedBroadcastChannel()`.

However with the `flow { }.stateIn(scope)` scenario I described in https://github.com/Kotlin/kotlinx.coroutines/issues/1973#issuecomment-621521467, I was expecting `stateIn` to have the following signature.

```kotlin
fun Flow<T>.stateIn(scope: CoroutineScope): Flow<T>
```

so there's no ability to actually read the current value. you just need to cache whatever you call `emit(..)` with

**Comment by zach-klippenstein** [comment]
> `fun Flow<T>.stateIn(scope: CoroutineScope): Flow<T>`

I like this API a lot more than the initially-proposed `suspend stateIn`: being able to access the "current state" is something I've only had to really do if I've got a `MutableStateFlow` and want to do something like `_state.value = _state.value.copy(…)`. I rarely if ever care about accessing the current value on a transformed stream.

> it would have a similar api like value (throws exception if not initialized) and `valueOrNull`

I don't really have any opinion about this, I would be happy with either. You can implement `LazyMutableStateFlow` on top of the current `MutableStateFlow`, and you can implement `EagerMutableStateFlow` on top of `LazyMutableStateFlow`, so I think it's just a question of which is the more common use case.

**Comment by lfmunoz** [comment]
Would a spreadsheet be a good use case for this? I'm thinking more of an admin panel that shows all the information status of a back-end service but it similar to a spreadsheet where variables propagate, or chain. I'm wondering if this would efficient to have a few 100 of these states? I don't suppose performance tests are done on this to give a rough idea of memory footprint?

**Comment by ZakTaccardi** [comment]
@zach-klippenstein 

> I don't really have any opinion about this, I would be happy with either.

I personally really like that `StateFlow` always has a value. It's a bit silly that `.valueOrNull` exists when you construct a `ConflatedBroadcastChannel(1)` - as its never `null`. `StateFlow<T>` can certainly be wrapped to provide `.valueOrNull` functionality.

> I rarely if ever care about accessing the current value on a transformed stream.

Agreed. If you need to read a `StateFlow<T>`'s current value at any time, `StateFlow<T>` or `LazyStateFlow<T>` would be more appropriate. Unless `stateIn(..)` provides an initial state, you cannot guarantee that there's a value available from the source flow.

```kotlin
val stateFlow: Flow<Int> = flow<Int> {
  // ..
}
  .stateIn(scope)
```

**Comment by LouisCAD** [comment]
Please, don't confuse `StateFlow` which has a goal of always having a value available, with Flow sharing, which is discussed in another issue #1261 (with a pending PR: #1716).

**Comment by ZakTaccardi** [comment]
> Please, don't confuse StateFlow which has a goal of always having a value available, with Flow sharing,

They're related. It makes sense that `.stateIn(scope)` would be a specific variant of how `.shareIn(..)` behaves. The term "state" has the implication that the latest value is replayed to new subscribers. And avoiding all the `.refCount(..)` and `autoConnect()` stuff from Rx would be amazing here at reducing complexity.

**Comment by zimmi** [comment]
I'd like to have a `StateFlow<List<DownloadProgress>>` that contains progress per file, with potentially multiple downloads active at any given time. I'd like to keep `DownloadProgress` mutable to avoid copying the `DownloadProgress` instance on every little bit of progress. This won't work with the current design however, because conflation is based on equality, requiring different instances. Can `StateFlow` help me with this?

One way would be to make conflation configurable, allowing to pass a function always returning false (i.e. turning conflation off). Maybe there is a better tool for this?

**Comment by Dico200** [comment]
@zimmi Considering `StateFlow` to be like `Observable`, you would need an `ObservableList` variant. It seems like an incorrect approach to have a mutable object in the StateFlow, effectively having to notify the StateFlow of changes by setting it again to the same reference.

**Comment by zimmi** [comment]
@Dico200 Thanks for the answer! I guess what I want is a `StateFlow` without baked in conflation. Adding it on top is easy, but removing it doesn't seem feasible without workarounds. If reasearch has shown that conflation is a useful assumption for most use cases then it's fine, I just wanted to share a scenario where that is not ideal. Replace "not wanting to allocate" with "can't create a new instance because it's a third party singleton / heavy resource handle that doesn't override equals" for a stronger case that I also encountered in the past.

More generally, what is the recommended way to tell listeners to look again at the last state? It might not be functionally pure, but I found this very useful in Apps I built in the past. I hope this is not too offtopic, but `StateFlow` seems to be designed to be a core building block of many Apps, I feel like there should be at least an answer to this question, if only in the docs.

**Comment by gajicm93** [comment]
@zimmi Totally agree with you, this case should be looked at and I used it many times also..

But a simple solution to avoid conflation problems would be to use a wrapper/holder object, for e.g:

```
data class Wrapper(
    val version: Int,
    val heavy: Object
)
```
And then every time you want to mutate a heavy Object, just wrap it into a new Wrapper object with increased version and you're done..

But this is a boilerplate workaround of course..

**Comment by Zhuinden** [comment]
Theoretically you could suspend when you invoke `value` if it was not initialized with a value, but I'm not sure if blocking silently in a CoroutineScope would be safer than either using an `Optional<T>`...

...wait, can't you just make it a `StateFlow<T?>` when you want the `value` to be nullable?

**Comment by ZakTaccardi** [comment]
> ...wait, can't you just make it a `StateFlow<T?>` when you want the `value` to be nullable? After all, while it is not initialized, the value is `absent`.


`null` can have semantic meaning. for example:

```
val currentUserId: Flow<UserId?>
```

1. lack of value: we don’t know yet. Info could still be loading from disk
2. `null`: no user found. We loaded from disk, and did not find an active user. Likely means that user is not logged in
3. non-null: logged in user

I don’t really want to start a discussion about how a `sealed class` should be used in this scenario to promote clarity, because yes I agree. But supporting this use case is still useful.

**Comment by ZakTaccardi** [comment]
I created a rudimentary [`NoInitialStateFlow`](https://gist.github.com/ZakTaccardi/dfe2679238b673ead610e9ed4bd8b49c). I'm certainly curious if my [`flow { .. }.conflate()`](https://gist.github.com/ZakTaccardi/dfe2679238b673ead610e9ed4bd8b49c#file-noinitialstateflowimpl-kt-L12-L17) is safe

**Comment by RobertKeazor** [comment]
hmm interesting. But it seems abit much, just to represent an empty state?  When you can just  emit a State.Empty/State.Loading/or some kind of flag that represents that represents the state not being initialized? What would be an example use case that would make one use this over just having a state representation of not being  initialized state?

**Comment by ZakTaccardi** [comment]
>  When you can just  emit a State.Empty/State.Loading/or some kind of flag that represents that represents the state not being initialized?

This increases verbosity. However, I do agree it’s a good practice.

> What would be an example use case that would make one use this over just having a state representation of not being  initialized state?

When you don’t want to define a sealed class for the following:

1. lack of value: we don’t know yet. Info could still be loading from disk
2. null: no user found. We loaded from disk, and did not find an active user. Likely means that user is not logged in
3. non-null: logged in user

It’s the same argument that allows nulls to be emitted in a `Flow<T?>` but not a `Publisher<T>`.

**Comment by elizarov** [comment]
There is a followup design on `SharedFlow` that closes most of the issues that were raised in the comments here. See #2034

**Comment by AdamSHurwitz** [comment]
It would desirable for the ability to create a `MutableStateFlow` with no value assigned to it similar to [`MutableLiveData`](https://developer.android.com/reference/androidx/lifecycle/MutableLiveData). 

I'm refactoring the Android ViewModel's view state and effects in the [CryptoTweets](https://github.com/AdamSHurwitz/CryptoTweets) sample app from `LiveData`/`MutableLiveData` to `StateFlow`/`MutableStateFlow` in order to update the view UI, in this case, a  fragment, accordingly. 

### LiveData

The `MutableLiveData` is created with no value assigned to it.

_FeedViewState.kt_

```Kotlin
data class _FeedViewState(
    val _feed: MutableLiveData<PagedList<Tweet>> = MutableLiveData()
)

data class FeedViewState(private val _viewState: _FeedViewState) {
    val feed: LiveData<PagedList<Tweet>> = _viewState._feed
}
```

### StateFlow

The `MutableStateFlow` is created with a null value, thus requiring the type contained by the `MutableStateFlow` to allow nulls. Ideally, it would be easier to create an empty `MutableStateFlow` because for this use case it would be difficult to create an empty PagedList to initialize the view state with.

_FeedViewState.kt_

```Kotlin
@ExperimentalCoroutinesApi
data class _FeedViewState (
    val _feed: MutableStateFlow<PagedList<Tweet>?> = MutableStateFlow(null),
)

@ExperimentalCoroutinesApi
data class FeedViewState(private val _viewState: _FeedViewState) {
    val feed: StateFlow<PagedList<Tweet>?> = _viewState._feed
}
```

In the ViewModel the view states are initialized with default empty values before being populated.

_FeedViewModel.kt_

```Kotlin
@ExperimentalCoroutinesApi
class FeedViewModel(private val feedRepository: FeedRepository) : ViewModel(), FeedViewEvent {
    val LOG_TAG = FeedViewModel::class.java.simpleName

    private val _viewState = _FeedViewState()
    val viewState = FeedViewState(_viewState)
   ...
}
```

**Comment by Zhuinden** [comment]
>      val _feed: MutableLiveData<PagedList<Tweet>> = MutableLiveData()

This doesn't do what you think it does

**Comment by AdamSHurwitz** [comment]
Thanks @Zhuinden. Is there a portion of the MutableLiveData code you'd recommend exploring in order to understand further?

**Comment by RobertKeazor** [comment]
I don't know, Im still in the belief that its better to have a state that represents missing item, rather than refactoring  it handle a missing value.  This api, seems to be built with the intention that 
value : T, trying to handle a missing a case to where a item is missing sounds more like a anti-pattern

**Comment by elizarov** [comment]
> _Please let me know if this is the best place for feedback on `StateFlow` since this issue has been closed._

@AdamSHurwitz It is better to create separate issues for new proposals.

> It would desirable for the ability to create a `MutableStateFlow` with no value assigned to it similar to [`MutableLiveData`]

This particular desire is taken into account by the design of `SharedFlow`. See #2034. Using `SharedFlow` you can have a `StateFlow`-like behavior without initial value and a number of other possible tweaks. If something is missing there for your use-case, please comment on the open issue #2034.

**Comment by Zhuinden** [comment]
>      interface MutableStateFlow<T> : StateFlow<T> {

Are you sure it's not an issue that you can potentially smart-cast an "immutable" state flow into a mutable one?

Of course, that's technically true of `List` and I haven't really seen people do it, just pass `MutableList`s around if they really want to by-pass "read-only"-ness.

**Comment by elizarov** [comment]
> Are you sure it's not an issue that you can potentially smart-cast an "immutable" state flow into a mutable one?

@Zhuinden We have quite lengthy experience with a similar design for `List`/`ModifiableList` in Kotlin and it is not causing any problems to be worried about it, as far as I know.

**Comment by Zhuinden** [comment]
Okay then, that should be great :slightly_smiling_face:

**Comment by muthuraj57** [comment]
> To support closing there would be a MutableStateFlow.close(cause: Throwable? = null): Boolean method for the owner of the state flow to close it via its writeable reference and StateFlow.isClosed: Boolean property on read-only interface.

Is this enhancement tracked elsewhere where I can follow? I'm trying to publish download progress and need to close the flow once the download completes so that the collector is finished.

**Comment by elizarov** [comment]
> Is this enhancement tracked elsewhere where I can follow? I'm trying to publish download progress and need to close the flow once the download completes so that the collector is finished.

@muthuraj57 No. We've abandoned it. On the collector side you can use `takeWhile { ... }` to stop collector when download completes. See longer discussion on this decision in #2034

**Comment by JonathanSampayoSolis** [comment]
This idea was huge good. I'm wondering if is still handled?
I'm facing an use case to send my repository states by reactive like ->**(ConflatedBroadcastChannel.send(Loading)**, **(ConflatedBroadcastChannel.send(Sucess)**, etc. But Conflated... is already obsolete. Any solution into channels lib?

**Comment by zach-klippenstein** [comment]
`StateFlow` _is_ the replacement for `ConflatedBroadcastChannel`.

**Comment by JonathanSampayoSolis** [comment]
> `StateFlow` _is_ the replacement for `ConflatedBroadcastChannel`.

Is it available for implementation already?

**Comment by zach-klippenstein** [comment]
It's not stable for implementing yourself, but it is stable for use in the latest coroutines release (see the kdoc).

**Comment by deinlandel** [comment]
Cannot find any actual implementation of asStateFlow() in source code, despite the fact that it is mentioned in the examples. It's also not recognized by the IDE

**Comment by gildor** [comment]
@deinlandel This function [called stateIn](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/state-in.html) in current implementation

**Comment by deinlandel** [comment]
@gildor  I indeed found this javadoc, but the problem is, I cannot locate this function in this repository's source code, and also my IDE fails to compile the code when I try to use it even with the latest kotlinx.coroutines libraries.

**Comment by LouisCAD** [comment]
@deinlandel I think there's a bug in GitHub search/indexing. Here's the permalink to the implementation of `asStateFlow()`: https://github.com/Kotlin/kotlinx.coroutines/blob/1b34e1c7dd6207d7683c307bae0b934a3dc18d09/kotlinx-coroutines-core/common/src/flow/operators/Share.kt#L357
